### DISCUSSION TITLE
Pickleable httpx.Request is not restored proper.

### DISCUSSION THREAD

**Discussion by cloasdata** [encode/httpx#2201] [Q&A]
```python

import pickle
import traceback
import httpx

class MyExtension:
    some_stuff = "stuff"


url = "https://github.com"
method = "GET"
extensions = {"a_proper_extension": MyExtension()}

request = httpx.Request(method, url, extensions=extensions)
print(request)

pickled_request = pickle.dumps(request)
print(pickled_request)

repickled_request = pickle.loads(pickled_request)
print(repickled_request)

try:
    assert repickled_request == request  # is not the same but should be equal
except AssertionError as e:
    print(traceback.format_exc())

print("extensions:", repickled_request.extensions) # gone! why

```

It turns out that httpx.Request can be pickled but loading the serialized byte code will not return the same object at all.
I did not a lot of research on it so far and I cannot tell were exactly the problem occurs in httpx.

Reason for pickling is, that I'm running the client in several child processes and this leads to strange errors like
`httpx.StreamClosed: Attempted to read or stream content, but the stream has been closed.`

At least some warning should be given, that this class is not safe to pickle even if you can pickle it.

**Comment by cloasdata** [comment]
Today I found some time to look through the source and found 
```python
    def __getstate__(self) -> typing.Dict[str, typing.Any]:
        return {
            name: value
            for name, value in self.__dict__.items()
            if name not in ["extensions", "stream"]
        }

    def __setstate__(self, state: typing.Dict[str, typing.Any]) -> None:
        for name, value in state.items():
            setattr(self, name, value)
        self.extensions = {}
        self.stream = UnattachedStream()
```
That explains my above problem. But I don't know, why one should remove this data from the object. Some explanation or warning would be nice...

**Comment by florimondmanca** [comment]
@cloasdata Hi, just a heads up that you might want to review https://github.com/encode/httpx/pull/1579 for background. It's still possible there are some developer experience niceties we could be adding...

**Comment by cloasdata** [comment]
To be honest I did not understand the background, why Request needs to be pickled exactly this way.

But a typical thing one may want to do is to put a fresh request on a queue and when it is time a worker may get this request and send it through the workers client. 
At least this was one of my assumption on the purpose of this class. Otherwise I could just use the client method get without having a stateful object created in front.

for now I wrote an simple adapter.
